File: python/valid_parenthesis.py
class Solution:
    def isValid(self, s: str) -> bool:
        brackets = []
        bracketMap = {
            "(": ")",
            "{": "}",
            "[": "]" 
        }
        
        for currBracket in s:
            isOpeningBracket = currBracket in bracketMap

            if isOpeningBracket:
                brackets.append(currBracket) 
            else:
                if len(brackets) == 0 or bracketMap[brackets.pop()] != currBracket:
                    return False

        # could be extra remaining closing brackets
        areNoBracketsLeft = len(brackets) == 0

        return areNoBracketsLeft


# test cases

# consider:
#   - when closing bracket does not match opening bracket
#   - when extra closing brackets at end (don't pop empty stack)
#   - when not enough closing brackets (remaining opening brackets in stack at end)

# (){}[]
# ({[]}[]())
        
# ([)]
# [ (  ]  [
        
# [( also invalid
# }}}}}}}}}}}}}}}} also invalid (edge case, pop only if len(stack) >= 1)

File: cpp/valid_parenthesis.cpp
class Solution {
public:
    // stack approach
    bool isValid(string s) {
        stack<char> openBrackets;
        unordered_map<char, char> bracketPairs = { {'(', ')'}, {'[', ']'}, {'{', '}'} };

        for (char &bracket : s) {
            // open bracket
            if (bracketPairs.find(bracket) != bracketPairs.end()) {
                openBrackets.push(bracket);
                continue;
            }
            // last open bracket should exist and match current closed bracket
            if (openBrackets.empty() || bracket != bracketPairs[openBrackets.top()]) {
                return false;
            }
            openBrackets.pop();
        }
        // check if not enough closed brackets for open brackets
        return openBrackets.empty();
    }
};
